### PR TITLE
test(multi-env): add function/bot e2e for multienv

### PR DESCRIPTION
**Pull request by a1exwang**
- add function/bot e2e with multi-env
- move test case to multienv folder
- add owners file

E2E test result: https://github.com/OfficeDev/TeamsFx/actions/runs/1244634201

https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_boards/board/t/Infrastructure%20Team/Stories

### CODE REVIEW

File: packages/cli/tests/commonlib/functionValidator.ts [hunk 1]
```diff
@@ -172,14 +187,16 @@ export class FunctionValidator {
...
 
-    console.log("Validating app service plan.");
-    const servicePlanResponse = await this.getWebappServicePlan(
-      this.subscriptionId,
-      this.rg,
-      functionObject.appServicePlanName,
-      token as string
-    );
-    chai.assert(servicePlanResponse, functionObject.appServicePlanName);
+    if (isMultiEnvEnabled) {
```
Comment (hund030) [inline packages/cli/tests/commonlib/functionValidator.ts:190]: I didn't remember in what situation we can't get app service name from config. Do we check app service plan when multi env enabled?

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        if (userdataKey in userdata) {
+          context[component][key] = userdata[key];
+        }
+      }
+    }
+  }
+  return ok(context);
+}
+
+describe("Create single tab/bot/function", function () {
```
Comment (hund030) [inline packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:69]: Seems the description doesn't match the behavior. You actually create a tab+bot+function.

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  const subscription = getSubscriptionId();
+  const projectPath = path.resolve(testFolder, appName);
+  const processEnv = mockTeamsfxMultiEnvFeatureFlag();
+
+  it(`Tab`, async function () {
+    // new a project (tab only)
+    try {
+      let result;
+      result = await execAsync(
+        `teamsfx new --interactive false --app-name ${appName} --capabilities tab bot --azure-resources function --programming-language javascript`,
```
[the remark this answers is not in the document]
Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:82]: I will add in the next pr

**Review by hund030 — APPROVED** [review]

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+describe("Create single tab/bot/function", function () {
+  const testFolder = getTestFolder();
+  const appName = getUniqueAppName();
+  const subscription = getSubscriptionId();
+  const projectPath = path.resolve(testFolder, appName);
+  const processEnv = mockTeamsfxMultiEnvFeatureFlag();
+
+  it(`Tab`, async function () {
+    // new a project (tab only)
```
[the remark this answers is not in the document]
Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts]: fixed

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+      }
+      ret[key] = obj;
+    } else {
+      ret[key] = value;
+    }
+  }
+  return ret;
+}
+
+// Load envProfile with userdata (not decrpted)
```
[the remark this answers is not in the document]
Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts]: fixed

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+      // add env
+      result = await execAsync(`teamsfx env add ${env} --env dev`, {
+        cwd: projectPath,
+        env: processEnv,
+        timeout: 0,
+      });
+      console.log(`[Successfully] env add, stdout: '${result.stdout}', stderr: '${result.stderr}'`);
+
+      // set active env
```
[the remark this answers is not in the document]
Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:114]: added

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts (hunk 2 above)
Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:70]: fixed

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        // Validate Function App
+        const func = FunctionValidator.init(context);
+        await FunctionValidator.validateProvision(func, false);
+
+        // Validate Bot Provision
+        const bot = BotValidator.init(context);
+        await BotValidator.validateProvision(bot);
+      }
+
+      // deploy
```
Comment (qinezh) [inline packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:189]: will the e2e test cases cover `publish` operation?

File: packages/cli/tests/commonlib/functionValidator.ts (hunk 1 above)
Comment (a1exwang) [reply packages/cli/tests/commonlib/functionValidator.ts:190]: 1. This condition check is wrong (it should be `if (!isMultiEnvEnabled)`, I will fix it)
2. In the multi-env + bicep case (these two flags should always be enabled together), the app service plan name is not in `profile.dev.json`, so we don't check app service plan in the multi-env case.

Example `profile.dev.json`:
Screenshot from 2021-09-17 16-40-48

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts (hunk 7 above)
Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:189]: Yes, it will be added in next few PRs.

**Review by qinezh — APPROVED** [review]

File: packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+      result = await execAsync(
+        `teamsfx new --interactive false --app-name ${appName} --capabilities tab bot --azure-resources function --programming-language javascript`,
+        {
+          cwd: testFolder,
+          env: processEnv,
+          timeout: 0,
+        }
+      );
+      console.log(
+        `[Successfully] scaffold to ${projectPath}, stdout: '${result.stdout}', stderr: '${result.stderr}''`
```
Comment (Alive-Fish) [inline packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:91]: I think stdout may be too much and hope not to print it if successfully.

Comment (a1exwang) [reply packages/cli/tests/e2e/multienv/TestRemoteHappyPath.tests.ts:91]: I think too much log is better than too little log. Especially for E2E, running a single test takes too much time. The effort of retrying the test and print the debug info is big. And I don't see why too much log will cause an issue here.

**Review by Alive-Fish — APPROVED** [review]

**Pull request merged**